src/client/datascience/jupyter/kernels/cellExecution.ts

```diff
@@ -316,32 +303,26 @@ export class CellExecution implements IDisposable {
         this.sendPerceivedCellExecute();
 
         traceCellMessage(this.cell, 'Update with error state & output');
-        // No need to append errors related to failures in Kernel execution in output.
-        // We will display messages for those.
+        // If the error doesn't derive from BaseError, it came from execution
         if (!(error instanceof BaseError)) {
             this.execution?.appendOutput([translateErrorOutput(createErrorOutput(error))]).then(noop, noop);
+        } else {
+            // Otherwise it's an error from the kernel itself. Put it into the cell
+            const failureInfo = analyzeKernelErrors(
+                workspace.workspaceFolders || [],
+                error,
+                getDisplayNameOrNameOfKernelConnection(this.kernelConnection),
+                this.kernelConnection.interpreter?.sysPrefix
+            );
+            if (failureInfo) {
+                this.execution
+                    ?.appendOutput([translateErrorOutput(createErrorOutputFromFailureInfo(failureInfo))])
```

Comment:
Previously error handler would deal with the errors, such as what to do where to display.
Now we have this class that displays the errors in the output & then error handler will display the errors as a popup message.

Can we centralize the code to displaying errors (in output & messagebox) in one place. In the past it was in a few different places and we centralized it to ensure we have a consnstent UX when things fail, and I'm afraid we'll start going down that path again.

**Another observation:**
Assume the kernel failure indicates a module could not be imported or something similar, and we need to perform additional analysis on the error (something that's async & possibly requires user feedback).
Once done, we'll have more information about the error & the context. Next we display the new error (after analysis) information in the output & we'd need to do the exact same thing all over again when displaying the popup error message in the error handler.

The idea behind the error handler was if something goes wrong, we handle it it one place. I know its still possible, but displaying it in two places now means we'd need to do some analysis in two places.  


If we can move it into a class or the like, that would be great.

Reply:
Note: This isn't a must, but would be great if we could achieve this some how, 
After all, this is a huge step forward in improving the code.